rockchip: add support for FriendlyARM NanoPi R2S

blocktrron opened the pull request:
## Hardware
RockChip RK3328 ARM64 (4 cores)
1GB DDR4 RAM
2x 1000 Base-T
3 LEDs (LAN / WAN / SYS)
1 Button (Reset)
Micro-SD slot
USB 2.0 Port

## Installation
Uncompress the OpenWrt sysupgrade and write it to a micro SD card using
dd.

## MAC-address
The vendor code supports reading a MAC address from an EEPROM connected
via i2c0 of the SoC. The EEPROM (address 0x51) should contain the MAC
address in binary at offset 0xfa. However, my two units didn't come with
such an EEPROM soldered on. The EEPROM should be placed between the SoC
and the GPIO pins on the board. (U10)

Generating rendom MAC addresses works around this issue. Otherwise, all
boards running the same image have identical MAC addresses.

lucize committed:
atf-rockchip: update to 2.3

also install the firmware for all the supported boards

Signed-off-by: Lucian Cristian <lucian.cristian@gmail.com>
[fix ATF blob path in uboot-rockchip]
Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
uboot-rockchip: update to v2020.07

Update the U-Boot to version v2020.07. Also replace the Makefile rewrite
with a proper patch, explaining why this hack is needed.

Run-tested: FriendlyARM NanoPi R2S

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
uboot-rockchip: add NanoPi R2S support

Add support for the FriendlyARM NanoPi R2S.

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
base-files: add function for generating random MAC

This adds a function for generating a valid random MAC address (unset MC
bit / set locally administered bit).

It is necessary for devices which do not have a MAC address programmed
by the manufacturer.

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
rockchip: allow for multiple boot-scripts

Add support for select a bootscript depending on the device built. This
is necessary, as the FriendlyARM NanoPi R2S needs a different bootcmd in
order to produce output on the debug UART.

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
rockchip: add NanoPi R2S support

Hardware
--------
RockChip RK3328 ARM64 (4 cores)
1GB DDR4 RAM
2x 1000 Base-T
3 LEDs (LAN / WAN / SYS)
1 Button (Reset)
Micro-SD slot
USB 2.0 Port

Installation
------------
Uncompress the OpenWrt sysupgrade and write it to a micro SD card using
dd.

MAC-address
-----------
The vendor code supports reading a MAC address from an EEPROM connected
via i2c0 of the SoC. The EEPROM (address 0x51) should contain the MAC
address in binary at offset 0xfa. However, my two units didn't come with
such an EEPROM soldered on. The EEPROM should be placed between the SoC
and the GPIO pins on the board. (U10)

Generating rendom MAC addresses works around this issue. Otherwise, all
boards running the same image have identical MAC addresses.

Signed-off-by: David Bauer <mail@david-bauer.net>

peterwillcn commented:
@blocktrron Fixed the LAN port indicator light is not working properly
@Jayanta525 patch fix the rtl8152 leds: https://github.com/jayanta525/openwrt-nanopi-r2s/blob/rk3328/target/linux/rockchip/patches-5.4/010-nanopi-r2s-rtl8152-leds.patch

blocktrron commented:
@peterwillcn 

> @jayanta525 Is this the same ugly hack FriendlyARM has in their tree? It would make much more sense to have the LED register configurable using OF. I don't want to carry such a hack for this functionality, even if it is comparably tiny, as it has to be maintained and is not upstreamable.
> 
> Also, the discussion here is not about the NanoPi R2S.

If someone upstreams OF support for the LED register, we can backport this. I don't see the ugly hack tracked downstream.

adrianschmutzler commented on package/boot/uboot-rockchip/Makefile:
```diff
@@ -27,19 +27,40 @@ define U-Boot/rockpro64-rk3399
+  ATF:=rk3399_bl31.elf
+endef
+
+define U-Boot/nanopi-r2s-rk3328
```
Consider moving that up due to alphabetic sorting (similarly for the UBOOT_TARGETS list below).

blocktrron replied:
I've sorted this by SoC now.

adrianschmutzler commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -174,6 +174,18 @@ macaddr_setbit_la() {
 	macaddr_setbit $1 7
 }
 
+macaddr_unsetbit_mc() {
```
I'd consider bumping PKG_RELEASE for base-files here, as what you add can be considered a feature ...

blocktrron replied:
Done.

adrianschmutzler commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -4,6 +4,17 @@
+  SOC := rk3328
+  UBOOT_DEVICE_NAME := nanopi-r2s-rk3328
+  IMAGE/sysupgrade.img.gz := boot-common | boot-script nanopi-r2s | pine64-img | gzip | append-metadata
+  DEVICE_PACKAGES := kmod-usb-net kmod-usb-net-rtl8152
```
kmod-usb-net is selected by kmod-usb-net-rtl8152, so could be removed

blocktrron replied:
It is not, it only has the dependency flagged without autoselection.

adrianschmutzler replied:
Oops, my eye was moving the plus to the wrong side of the equal sign. Sorry.

I will think about whether that should be made a +kmod-usb-net in AddDepends/usb-net separately ...

peterwillcn commented on target/linux/rockchip/patches-5.4/001-rockchip-rk3328-Add-support-for-FriendlyARM-NanoPi-R.patch:
```diff
@@ -0,0 +1,347 @@
++	rockchip,hw-tshut-polarity = <0>;
++	status = "okay";
++};
++
```
It may be possible to turn on UART1, such as PPS NTP Server using this base port ?
/bin/stty -F /dev/ttyS1 9600
```
&uart1 {
        current-speed = <9600>;
        status = "okay";
};

pps {
         pinctrl-names = "default";
         compatible = "pps-gpio";
         gpios = <&gpio2 RK_PA2 GPIO_ACTIVE_HIGH>;
         assert-falling-edge;
};
```

kongfl888 commented:
Good job! 👍

mj22226 commented:
I have an issue on this PR, when the device boot with the lan cable unplugged, it don't get LAN IP as I'm using a USB wireless dongle to connect, It work fine if I change LAN on eth0 , WAN  on eth1.

blocktrron commented:
@mj22226 I can't reproduce the behavior you've outlined. For me both interfaces work regardless of the link state on boot.

mj22226 commented:
@blocktrron Well, perhaps I have a different board revision:

![Screenshot from 2020-07-25 14-00-16](https://user-images.githubusercontent.com/67804477/88463231-4e6a9e00-ce7f-11ea-9bf7-37c0d81650ea.png)

blocktrron commented:
My device also is a v1, so this should not make a difference.

mj22226 commented:
@blocktrron Possibly is just my board, other than the LAN issue it is working great. Thank you.

blocktrron committed:
rockchip: enable LAN LED configuration

This enables the LEDs on the LAN interfaces.

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron commented:
Added a patch to configure the USB ethernet adapters LED via OF and rebased on master.

EnfermeraSexy commented:
I can't change sys led (the red one), it always stays solid whatever setting you try

blocktrron commented:
@EnfermeraSexy this might be a bug in the OpenWrt LED handling, does changing the brightness via sysfs work?

EnfermeraSexy commented:
> @EnfermeraSexy this might be a bug in the OpenWrt LED handling not specific to the board. Does changing the brightness via sysfs work?

Yes, it works with the sysfs settings the bridgnesth to 0.

> There seems to be a bug in LuCI LED handling. Set the default option for the sys LED to 0 and the trigger to none. This should persistently turn of the LED. But again, this is not board specific.

Seems to be a problem with luci true. Setting manually the **default** option makes the led to turn off. With the patches from jayanta525 it wasnt needed to set that option to 0.

Another thing that im missing from jayanta525 patches is the thermal sensor, wich i think is solved with this patch
https://github.com/jayanta525/openwrt-nanopi-r2s/blob/rk3328/target/linux/rockchip/patches-5.4/005-add_hwmon_support.patch

blocktrron commented:
@EnfermeraSexy The patch you are referring to is a hack to expose the temperature to sysfs. I'm not really in favor of carrying such non-upstreamable hacks. If someone implements this in a ways it's accepted upstream, we can backport this.

Regarding the LEDs: I've added a fix for the LED handling in my staging tree which I'd like to commit seperately from this board. https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=summary

peterwillcn commented:
IRQ eth_wake_irq not found
```
[    0.338804] rk_gmac-dwmac ff540000.ethernet: IRQ eth_wake_irq not found
[    0.339399] rk_gmac-dwmac ff540000.ethernet: IRQ eth_lpi not found
```

blocktrron commented:
@peterwillcn Can you please explain the problem with this? Both IRQs are optional and their availability depend on the platform.

1715173329 commented:
Urr... Any one has tested the NAT throughput speed on ETH1? I can only reach to 3xx Mbps with the source code of this repo.
When I use @jayanta525's, it performs well.
![image](https://user-images.githubusercontent.com/22235437/88531444-70a61c80-d035-11ea-9a02-7c1640916d83.png)
![image](https://user-images.githubusercontent.com/22235437/88531453-77cd2a80-d035-11ea-8075-03b5ca5e36b2.png)

blocktrron commented:
@1715173329 I have no idea what is going on in his repo, but with irqbalance installed and enabled, I'm able to get 1 Gbit/s on LAN and WAN ports in both directions.

1715173329 commented:
> @1715173329 I have no idea what is going on in his repo, but with irqbalance installed and enabled, I'm able to get 1 Gbit/s on LAN and WAN ports in both directions.

Yep, me too. I don't even know what those patches in his repo did (so ugly).
And, I noticed that some patches had changed since your first commit in your private repo, I will give it a new try with recently changes.
BTW, irqbalance seemed has some issues when processing traffic from ETH1, so I wrote a script instead.

peterwillcn commented:
@blocktrron 
```
root@OpenWrt:~# openssl speed aes-128-cbc
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
aes-128 cbc      39738.11k    48405.72k    51064.13k    51783.50k    51830.49k    52001.66k
```
When I use @jayanta525's, it openssl performs well.
```
root@OpenWrt:~# openssl speed aes-128-cbc
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
aes-128 cbc      55115.86k    60549.91k    62123.52k    62787.30k    62698.84k    62701.57k

1715173329 commented:
> When I use @jayanta525's, it openssl performs well.

Maybe u need selete the kernel crypto modules.

blocktrron commented:
@1715173329 ARM64 crypto extensions are not yet packaged in OpenWrt. So it has to be enabled for the kernel unless someone packages it.

But again, this is not specific to the board and should be kept separate. Feel free to open a PR. Most likely, the kernel partition has to be resized as we are using an uncompressed image (arm64 does not support zImage yet).

peterwillcn commented:
we should optimize hardware performance to the limit rather than think about some limitations ;-）

1715173329 commented:
> we should optimize hardware performance to the limit rather than think about some limitations ;-）

Yes, that's we actually did.

1715173329 commented:
@blocktrron irqbalance will process all the traffic via cpu0, and this will cause speed down.
![image](https://user-images.githubusercontent.com/22235437/88558192-cd1d3200-d05d-11ea-93ba-61bea3527846.png)

blocktrron commented:
@1715173329 yup, this is the way I've tested this. Anyway, I'm also not in favor of adding `irqbalance` to the default package set, as this seems overkill and introduces more configuration state.

Can you share the script you've wrote? I' would like to do it in a way it acquires the interrupts to distribute from OF. The stuff I've seen until now relies on hardcoded paths.

blocktrron commented:
@peterwillcn That's out of question (it was rather a reminder to the person taking care of this). Yet, this has nothing to do with the board, as this affects the whole platform (and should be done separately).

1715173329 commented:
@blocktrron Here is the script: https://paste.ubuntu.com/p/tXf8XTWJHW/
I'll keep in finding the issue with the NAT thoughtput.

peterwillcn commented:
> @blocktrron Here is the script: https://paste.ubuntu.com/p/tXf8XTWJHW/ (maybe need some adjustment
> I'll keep in finding the issue with the NAT thoughtput.

![Screen Shot 2020-07-27 at 23 34 44](https://user-images.githubusercontent.com/323794/88561306-ce505e00-d061-11ea-8d8a-9ba7fed32b5e.png)

1715173329 commented:
@peterwillcn This is the result I got. (Cuased by the NAT thoughtput issue, RX can only reach to 3xx Mbps)
![image](https://user-images.githubusercontent.com/22235437/88561563-29825080-d062-11ea-8c32-70aee8e37b5c.png)

blocktrron committed:
arm64: dts: rockchip: Add txpbl node for RK3399/RK3328

Some rockchip SoCs like the RK3399 and RK3328 exhibit an issue
where tx checksumming does not work with packets larger than 1498.

The default Programmable Buffer Length for TX in these GMAC's is
not suitable for MTUs higher than 1498. The workaround is to disable
TX offloading with 'ethtool -K eth0 tx off rx off' causing performance
impacts as it disables hardware checksumming.

This patch sets snps,txpbl to 0x4 which is a safe number tested ok for
the most popular MTU value of 1500.

For reference, see https://lkml.org/lkml/2019/4/1/1382.

Signed-off-by: Carlos de Paula <me@carlosedp.com>
Link: https://lore.kernel.org/r/20200218221040.10955-1-me@carlosedp.com
Signed-off-by: Heiko Stuebner <heiko@sntech.de>
Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron commented:
@1715173329 please try the latest branch. Regarding the IRQ steering ,it seems like the packet steering mechanism is to limited for our needs, as it does fail to detect the IRQs for both interfaces. I'll look into creating a intermediate target-specific solution.

1715173329 commented:
@blocktrron Thanks for your hard work!
Howerver now it's 4 A.M. in my country, so I have to go sleep, do sorry for that.
I'll give a report in the daytime.

blocktrron committed:
rockchip: distribute net interrupts

This adds a hotplug script for distributing interrupts of eth0 and eth1
across different cores. Otherwise the forwarding performance between
eth0 and eth1 is severely affected.

The existing SMP distribution mechanic in OpenWrt can't be used here, as
the actual device IRQ has to be moved to dedicated cores. In case of
eth1, this is in fact the USB3 controller.

Signed-off-by: David Bauer <mail@david-bauer.net>

peterwillcn commented:
@blocktrron  latest branch
![Screen Shot 2020-07-28 at 12 27 43](https://user-images.githubusercontent.com/323794/88619288-cdedac80-d0cd-11ea-8c2d-4ddf1d508a17.png)

1715173329 commented:
@blocktrron It performs perfect now.
And your IRQ script works fine, cpu1 will process eth0 and cpu2 will process eth1.
![image](https://user-images.githubusercontent.com/22235437/88631035-187a2380-d0e4-11ea-8724-acfe58457a53.png)

blocktrron commented:
I'll merge this then. For any further performance enhancements regarding the rockchip platform, feel free to open a PR. @Manawyrm has probably more knowledge in that regard.
